### ISSUE TITLE
[4][gibbsfromncis] First prototype visual concept

### ISSUE THREAD

**Issue by wmbutler** [bitshares/bitshares-pay#1]
Image on the left is the bitpay presentation for payment. On right is the bitshares MVP concept

<img width="763" alt="screen shot 2018-01-08 at 12 34 41 am" src="https://user-images.githubusercontent.com/1254810/34660977-0712822a-f40c-11e7-8a38-c1eb4573b762.png">

**Comment by oxarbitrage** [comment]
this looks good @wmbutler , if we can make it work this will be big. if you need any help with the project let me know.

**Comment by xeroc** [comment]
Nice concept. Can I request that we figure out a "nicer" (read: shorter) string to put into the QR code because, the more content you put into QR codes, the bigger they become and more difficult to read through the cam.

Something like this:

    1;0;operation-payload
    | | |- payload: transfer=asset-id,amount,memo
    | |- type: 0=transfer
    |- version

**Comment by wmbutler** [comment]
Regarding the length of a QR code: (that's me being a smartass)

<img width="229" alt="screen shot 2018-01-08 at 6 21 57 pm" src="https://user-images.githubusercontent.com/1254810/34699028-dc464294-f4a0-11e7-8ed5-45936eb02b75.png">

**Comment by wmbutler** [comment]
@xeroc I'm trying to roughly follow how it's done with bitcoin and similar coins now:

https://bitcore.io/api/lib/uri

**Comment by wmbutler** [comment]
> this looks good @wmbutler , if we can make it work this will be big. if you need any help with the project let me know.

I think the major issues I'm thinking through now include:

- Polling the blockchain while waiting for the payment so we can provide feedback to the user indicating that the payment was received.
- Determining how to know the correct payment was received for a given transaction.
- Possibly including the private memo key as a variable for the vendor so they could perform a match on the memo (invoice number etc.)

If we don't utilize the memo matching option, we have a somewhat crude option of asking the purchaser to enter their `from` account name prior to the transaction. Then, we at least can match on: `to`, `from`, `amount`, `asset`.

Better than this would be to match in `to`, `amount`, `memo`, `asset`. But then we are relying on the user to either enter the memo properly or on all BTS compliant wallets or we are relying on QR codes to be properly interprested by BTS compliant wallets.

We need to make this as easy as possible without creating headaches. Ideally, non-conformant transactions could be refunded automatically with a reason in the memo response. This would provide a zero maintenance approach for users, but would also require that the private key for sending transactions be exposed on some level. I suppose the risk here could be mitigated by encouraging payment processors to sweep the deposit account on a frequent basis.

**Comment by abitmore** [comment]
@xeroc We should make a consensus/standard/protocol about QR codes and document it somewhere, like discussed in https://github.com/bitshares/bitshares-ui/issues/775#issuecomment-350202771. Actually MagicWallet has implemented the scanning feature according to consensus made in that issue, so better just follow it if no major issue found.

**Comment by abitmore** [comment]
In short, make it compatible with every client, otherwise it would be a mess.

**Comment by wmbutler** [comment]
I agree. This seems like something for the foundation. Can you get something published and linked from the .org website so we can move forward?

**Comment by gibbsfromncis** [comment]
@wmbutler I can help with UI for this. So just mention me when you need help.

**Comment by wmbutler** [comment]
Do it!! Please get Dominick's feedback as well.

**Comment by gibbsfromncis** [comment]
@wmbutler who is Dominick?

**Comment by gibbsfromncis** [comment]
@wmbutler lets discuss specific features you want me to delivery. I suppose for now I can delivery the following things: 
- [ ] Setup front-end infrastructure for future bitshares-pay (react, webpack, scss support etc.)
- [ ] Setup some page on the app like: https://pay.bitshares.org/payment?asset_id=1.3.121&amount=100&order=1234567890&receiver=bitshares-ui
- [ ] Fetch bitshares transactions while not found transaction with all fields matched requirements (to, memo, amount, asset_id). If transaction found I would display some "Success message". But user will see "success" message only if he pay after he open the page because parsing of past transactions log is not a good idea. I'll parse only new logs which come to me from blockchain.

If you agree I can start. But it would be much more than 4 hours. I suppose this requires around 12-15 hours.

**Comment by gibbsfromncis** [comment]
@wmbutler if you want me to create this feature based on of bitshares-ui repo it can be faster.

**Comment by gibbsfromncis** [comment]
@wmbutler ping

**Comment by wmbutler** [comment]
@gibbsfromncis This will be a freestanding project I think because others may want to incorporate the logic into their wallets. The UI may want to incorporate it as well.

I think it's a mistake to force the user to resolve back to a particular address. are better off with something like:

https://github.com/bitshares/bitshares-ui/issues/775#issuecomment-350202771

```
bts:bitshares-ui?amount=100000&asset=1.2.5&memo="This is the memo"
```

```
bts:<account>?amount=<amount>&asset=<asset_id>&memo="<memo>"
```

I'm not a fan of the URL without specific named parameters as I think this could lead to confusion long term.

Dominic = @ahdigital

**Comment by gibbsfromncis** [comment]
@wmbutler I understand your point about unified urls. But I'm talking about not unified url which we will use to scan on barcode or like a link to open pay page on mobile clients. I'm talking about to make simple test page: "https://pay.bitshares.org/payment?asset_id=1.3.121&amount=100&order=1234567890&receiver=bitshares-ui". On future it can be changed for any needs. But for now it would be good step to test it.

**Comment by wmbutler** [comment]
@gibbsfromncis here is the problem as I see it. You need a URI that calls an app or a web page. This is currently being done in the crypto world not via http but by using the asset name to call whichever application the user happens to choose as their wallet.

This page really explains it well:
https://bitcore.io/api/lib/uri
```
bitcoin:12A1MyfXbW6RhdRAZEqofac5jCQQjwEPBu
bitcoin:12A1MyfXbW6RhdRAZEqofac5jCQQjwEPBu?amount=1.2
bitcoin:12A1MyfXbW6RhdRAZEqofac5jCQQjwEPBu?amount=1.2&message=Payment&label=Satoshi&extra=other-param
```

The required piece is the address and everything else is optional and are denoted as such by params. The same should hold true for bitshares. The URI is very descriptive and can be digested by anyone without even looking at a key. This is why I'm opposed to the shorthand method proposed at bitshares/bitshares-ui#775 (comment). I am posting there next to further clarify my thoughts.

**Comment by wmbutler** [comment]
Restating here:

```
bitshares:<account>?asset=<asset>&amount=<amount>&memo=<memo>
```

| part | definition |
|------|-----------|
| bitshares | URL schema, all BitShares client use it. |
| account |  path part 3, target account name or id (1.2.xxx) . mandatory |
| asset      |  path part 4, asset symbol abbreviation or asset_id, like 'BTS' or 1.x.x.  according to BitShares, upper or lower case. mandatory        |
| amount  |  path part 5, amount, float number. optional. full precision of asset |
| memo    |   path part 6, memo.optional.   |
